File: javascript/src/fedml/cross_web/client/fedml_client_master_manager.ts
import { FedMLCommManager } from '../../core/distributed/fedml_comm_manager';
import { Message } from '../../core/distributed/communication/message';
import { event } from '../../mlops/mlops_init';
import { MyMessage } from './message_define';
import { TrainerDistAdapter } from './fedml_trainer_dist_adapter';
// import { computed } from 'vue';
// import { useAppStore } from '/@/store/modules/app';

// const appStore = useAppStore();
// const currClient = computed(() => appStore.getClient);
export class ClientMasterManager extends FedMLCommManager {
  args: object;
  trainer_dist_adapter: TrainerDistAdapter;
  num_rounds: number;
  round_idx: number;
  rank;
  client_real_ids;
  client_real_id;
  has_sent_online_msg;
  message_handler_dict;
  is_inited;

  constructor(
    args,
    trainer_dist_adapter: TrainerDistAdapter,
    comm = null,
    rank = 0,
    size = 0,
    backend = 'MPI',
  ) {
    super(args, comm, rank, size, backend);

    this.trainer_dist_adapter = trainer_dist_adapter;
    this.args = args;
    this.num_rounds = args.comm_round;
    this.round_idx = 0;
    this.rank = rank;
    this.client_real_ids = args.client_id_list;
    this.message_handler_dict = {};
    this.client_real_id = this.client_real_ids[0];
    this.has_sent_online_msg = false;
    this.is_inited = false;
  }

  register_message_receive_handlers() {
    console.log('fedml_runner register_message_receive_handlers');
    super.register_message_receive_handler(
      MyMessage.MSG_TYPE_CONNECTION_IS_READY,
      this.handle_message_connection_ready.bind(this),
    );
    super.register_message_receive_handler(
      MyMessage.MSG_TYPE_S2C_CHECK_CLIENT_STATUS,
      this.handle_message_check_status.bind(this),
    );
    super.register_message_receive_handler(
      MyMessage.MSG_TYPE_S2C_INIT_CONFIG,
      this.handle_message_init.bind(this),
    );
    super.register_message_receive_handler(
      MyMessage.MSG_TYPE_S2C_SYNC_MODEL_TO_CLIENT,
      this.handle_message_receive_model_from_server.bind(this),
    );
    super.register_message_receive_handler(
      MyMessage.MSG_TYPE_S2C_FINISH,
      this.handle_message_finish.bind(this),
    );
  }

  handle_message_connection_ready(msg) {
    console.log('handle_message_connection_ready ', msg);
    if (!this.has_sent_online_msg) {
      this.has_sent_online_msg = true;
      this.send_client_status(0);
    }
  }

  handle_message_check_status(msg) {
    console.log('handle_message_check_status ', msg);
    this.send_client_status(0);
  }

  async handle_message_init(msg) {
    if (this.is_inited) {
      return;
    }

    this.is_inited = true;
    console.log('handle_message_init: ', msg);
    const global_model_params = msg.msg_params[MyMessage.MSG_ARG_KEY_MODEL_PARAMS];
    const data_silo_index = msg.msg_params[MyMessage.MSG_ARG_KEY_CLIENT_INDEX];
    console.log('data_silo_index = ', data_silo_index);
    console.log('handle_message_init ', global_model_params);
    this.report_client_training_status(
      this.args.client_id_list[0],
      this.args.run_id,
      MyMessage.MSG_MLOPS_CLIENT_STATUS_TRAINING,
    );
    this.trainer_dist_adapter.update_model(global_model_params);
    this.round_idx = 0;
    await this.train();
    this.round_idx += 1;
  }

  async handle_message_receive_model_from_server(msg) {
    console.log('handle_message_receive_model_from_server.', msg);
    console.log('check the args-content: ', this.args);
    const model_params = msg.msg_params.model_params;
    console.log('handle_message_receive_model_from_server ', model_params);
    this.trainer_dist_adapter.update_model(model_params);

    if (this.round_idx < this.num_rounds) {
      await this.train();
      this.round_idx += 1;
    } else {
      this.send_client_status(0, MyMessage.MSG_MLOPS_CLIENT_STATUS_FINISHED);
      this.report_client_training_status(
        this.args.client_id_list[0],
        this.args.run_id,
        MyMessage.MSG_MLOPS_CLIENT_STATUS_FINISHED,
      );
      super.finish();
    }
  }

  handle_message_finish(msg) {
    console.log('====================cleanup ====================');
    console.log('handle_message_finish.', msg);
    this.cleanup();
  }

  cleanup() {
    this.send_client_status(0, MyMessage.MSG_MLOPS_SERVER_STATUS_FINISHED);
    this.report_client_training_status(
      this.args.client_id_list[0],
      this.args.run_id,
      MyMessage.MSG_MLOPS_SERVER_STATUS_FINISHED,
    );
    super.finish();
    // currClient.value.end();
  }

  report_client_training_status(edge_id, runId, status) {
    console.log('report_client_training_status ', runId);

    this.common_report_client_training_status(edge_id, runId, status);

    this.common_report_client_id_status(edge_id, runId, status);

    this.report_client_device_status_to_web_ui(edge_id, runId, status);
  }

  broadcast_client_training_status(edge_id, runId, status) {
    this.common_report_client_training_status(edge_id, runId, status);

    this.report_client_device_status_to_web_ui(edge_id, runId, status);
  }

  report_client_device_status_to_web_ui(edge_id, runId, status) {
    console.log('report_client_device_status_to_web_ui');
    const run_id = runId;
    const payload = {
      run_id: run_id,
      edge_id: edge_id,
      status: status,
      version: 'v1.0',
    };
    this.com_manager.mqtt_mgr.publish(
      'fl_client/mlops/status',
      JSON.stringify(payload),
      { qos: 2, retain: true },
      function (error) {
        if (error) {
          console.log(error);
        } else {
          console.log('report_client_device_status_to_web_ui Published');
        }
      },
    );
  }

  common_report_client_id_status(edge_id, runId, status) {
    console.log('report_client_id_status');
    const run_id = runId;
    const payload = {
      run_id: run_id,
      edge_id: edge_id,
      status: status,
    };
    this.com_manager.mqtt_mgr.publish(
      'fl_client/flclient_agent_' + edge_id + '/status',
      JSON.stringify(payload),
      { qos: 2, retain: true },
      function (error) {
        if (error) {
          console.log(error);
        } else {
          console.log('report_client_id_status Published');
        }
      },
    );
  }

  common_report_client_training_status(edge_id, runId, status) {
    const run_id = runId;
    const payload = {
      run_id: run_id,
      edge_id: edge_id,
      status: status,
    };
    this.com_manager.mqtt_mgr.publish(
      'fl_run/fl_client/mlops/status',
      JSON.stringify(payload),
      { qos: 2, retain: true },
      function (error) {
        if (error) {
          console.log(error);
        } else {
          console.log('common_report_client_training_status Published');
        }
      },
    );
  }

  async send_model_to_server(receive_id, weights) {
    event('comm_c2s', true, String(this.round_idx));
    const message = new Message(
      MyMessage.MSG_TYPE_C2S_SEND_MODEL_TO_SERVER,
      this.client_real_id,
      receive_id,
    );
    console.log('send_model_to_server');
    message.add_params(MyMessage.MSG_ARG_KEY_MODEL_PARAMS, weights);
    message.add_params(MyMessage.MSG_ARG_KEY_NUM_SAMPLES, 10);
    message.add_params(MyMessage.MSG_ARG_KEY_CLIENT_OS, 'web');
    const model_url = await this.send_message(message);
    console.log('check the model_url: ', model_url);
    await this.log_client_model_info(this.round_idx + 1, model_url);
  }

  async send_client_status(receive_id, status = 'ONLINE') {
    console.log('send_client_status ', status);
    const message = new Message(
      MyMessage.MSG_TYPE_C2S_CLIENT_STATUS,
      this.client_real_id,
      receive_id,
    );
    message.add_params(MyMessage.MSG_ARG_KEY_CLIENT_STATUS, status);
    message.add_params(MyMessage.MSG_ARG_KEY_CLIENT_OS, 'web');
    await this.send_message(message);
  }

  async log_client_model_info(round_index, model_url) {
    if (model_url == 'undefined') {
      return;
    }
    console.log('model info ', model_url);
    const model_info = {
      run_id: this.args.run_id,
      edge_id: this.args.client_id_list[0],
      round_idx: round_index,
      client_model_s3_address: model_url,
    };
    const message_json = JSON.stringify(model_info);
    this.com_manager.mqtt_mgr.publish(
      'fl_server/mlops/client_model',
      message_json,
      { qos: 2, retain: true },
      function (error) {
        if (error) {
          console.log(error);
        } else {
          console.log('Client Model Published');
        }
      },
    );
  }

  async train() {
    event('train', true, String(this.round_idx));

    const weights = await this.trainer_dist_adapter.train(this.round_idx);

    event('train', false, String(this.round_idx));

    await this.send_model_to_server(0, weights);
  }

  run(): void {
    this.register_message_receive_handlers();
  }
}

export default ClientMasterManager;

File: python/fedml/serving/client/fedml_client_master_manager.py
import json
import logging
import platform
import time

import torch.distributed as dist

from fedml import mlops
from fedml.constants import FEDML_CROSS_SILO_SCENARIO_HIERARCHICAL
from .message_define import MyMessage
from .utils import convert_model_params_from_ddp, convert_model_params_to_ddp
from ...core.distributed.fedml_comm_manager import FedMLCommManager
from ...core.distributed.communication.message import Message
from ...core.mlops.mlops_profiler_event import MLOpsProfilerEvent


class ClientMasterManager(FedMLCommManager):
    ONLINE_STATUS_FLAG = "ONLINE"
    RUN_FINISHED_STATUS_FLAG = "FINISHED"

    def __init__(self, args, trainer_dist_adapter, comm=None, rank=0, size=0, backend="MPI"):
        """
        Initialize the ClientMasterManager.

        Args:
            args: Arguments and configuration for the client manager.
            trainer_dist_adapter: Trainer distribution adapter for distributed training.
            comm: Communication backend (MPI, etc.).
            rank: Rank of the client.
            size: Size of the client group.
            backend: Backend for distributed training (MPI, etc.).
        """
        super().__init__(args, comm, rank, size, backend)
        self.trainer_dist_adapter = trainer_dist_adapter
        self.args = args

        self.num_rounds = args.comm_round
        self.round_idx = 0
        self.rank = rank
        self.client_real_ids = json.loads(args.client_id_list)
        logging.info("self.client_real_ids = {}".format(self.client_real_ids))
        # for the client, len(self.client_real_ids)==1: we only specify its client id in the list, not including others.
        self.client_real_id = self.client_real_ids[0]

        self.has_sent_online_msg = False
        self.is_inited = False

    def register_message_receive_handlers(self):
        """
        Register message receive handlers for handling various types of messages.
        """
        self.register_message_receive_handler(
            MyMessage.MSG_TYPE_CONNECTION_IS_READY, self.handle_message_connection_ready
        )

        self.register_message_receive_handler(
            MyMessage.MSG_TYPE_S2C_CHECK_CLIENT_STATUS, self.handle_message_check_status
        )

        self.register_message_receive_handler(MyMessage.MSG_TYPE_S2C_INIT_CONFIG, self.handle_message_init)
        self.register_message_receive_handler(
            MyMessage.MSG_TYPE_S2C_SYNC_MODEL_TO_CLIENT, self.handle_message_receive_model_from_server,
        )

        self.register_message_receive_handler(
            MyMessage.MSG_TYPE_S2C_FINISH, self.handle_message_finish,
        )

    def handle_message_connection_ready(self, msg_params):
        """
        Handle the "connection ready" message.

        Args:
            msg_params: Parameters of the message.
        """
        if not self.has_sent_online_msg:
            self.has_sent_online_msg = True
            self.send_client_status(0)

            mlops.log_sys_perf(self.args)

    def handle_message_check_status(self, msg_params):
        """
        Handle the "check client status" message.

        Args:
            msg_params: Parameters of the message.
        """
        self.send_client_status(0)

    def handle_message_init(self, msg_params):
        """
        Handle the "initialize" message and prepare for training.

        Args:
            msg_params: Parameters of the message.
        """
        if self.is_inited:
            return

        self.is_inited = True

        global_model_params = msg_params.get(MyMessage.MSG_ARG_KEY_MODEL_PARAMS)
        data_silo_index = msg_params.get(MyMessage.MSG_ARG_KEY_CLIENT_INDEX)

        logging.info("data_silo_index = %s" % str(data_silo_index))

        # Notify MLOps with training status.
        self.report_training_status(MyMessage.MSG_MLOPS_CLIENT_STATUS_TRAINING)

        if self.args.scenario == FEDML_CROSS_SILO_SCENARIO_HIERARCHICAL:
            global_model_params = convert_model_params_to_ddp(global_model_params)
            self.sync_process_group(0, global_model_params, data_silo_index)

        self.trainer_dist_adapter.update_dataset(int(data_silo_index))
        self.trainer_dist_adapter.update_model(global_model_params)
        self.round_idx = 0

        self.__train()
        self.round_idx += 1

    def handle_message_receive_model_from_server(self, msg_params):
        """
        Handle the "receive model from server" message.

        Args:
            msg_params: Parameters of the message.
        """
        logging.info("handle_message_receive_model_from_server.")
        model_params = msg_params.get(MyMessage.MSG_ARG_KEY_MODEL_PARAMS)
        client_index = msg_params.get(MyMessage.MSG_ARG_KEY_CLIENT_INDEX)

        if self.args.scenario == FEDML_CROSS_SILO_SCENARIO_HIERARCHICAL:
            model_params = convert_model_params_to_ddp(model_params)
            self.sync_process_group(self.round_idx, model_params, client_index)

        self.trainer_dist_adapter.update_dataset(int(client_index))
        logging.info("current round index {}, total rounds {}".format(self.round_idx, self.num_rounds))
        self.trainer_dist_adapter.update_model(model_params)
        if self.round_idx < self.num_rounds:
            self.__train()
            self.round_idx += 1
        else:
            self.send_client_status(0, ClientMasterManager.RUN_FINISHED_STATUS_FLAG)
            mlops.log_training_finished_status()
            self.finish()

    def handle_message_finish(self, msg_params):
        """
        Handle the "finish" message and perform cleanup.

        Args:
            msg_params: Parameters of the message.
        """
        logging.info(" ====================cleanup ====================")
        self.cleanup()

    def cleanup(self):
        """
        Perform cleanup operations at the end of training.
        """
        self.send_client_status(0, ClientMasterManager.RUN_FINISHED_STATUS_FLAG)
        mlops.log_training_finished_status()
        self.finish()

    def send_model_to_server(self, receive_id, weights, local_sample_num):
        """
        Send the model to the server.

        Args:
            receive_id: ID of the recipient (usually the server).
            weights: Model weights to be sent.
            local_sample_num: Number of local training samples.

        Note:
            This method sends model parameters to the server for aggregation.
        """
        tick = time.time()
        mlops.event("comm_c2s", event_started=True, event_value=str(self.round_idx))
        message = Message(MyMessage.MSG_TYPE_C2S_SEND_MODEL_TO_SERVER, self.client_real_id, receive_id,)
        message.add_params(MyMessage.MSG_ARG_KEY_MODEL_PARAMS, weights)
        message.add_params(MyMessage.MSG_ARG_KEY_NUM_SAMPLES, local_sample_num)
        self.send_message(message)

        MLOpsProfilerEvent.log_to_wandb({"Communication/Send_Total": time.time() - tick})
        mlops.log_client_model_info(
            self.round_idx+1, self.num_rounds, model_url=message.get(MyMessage.MSG_ARG_KEY_MODEL_PARAMS_URL),
        )

    def send_client_status(self, receive_id, status=ONLINE_STATUS_FLAG):
        """
        Send the client status message to the specified recipient.

        Args:
            receive_id: ID of the recipient.
            status: Status flag to be sent (default is ONLINE_STATUS_FLAG).

        Note:
            This method sends information about the client's status, including the operating system.

        """
        logging.info("send_client_status")
        logging.info("self.client_real_id = {}".format(self.client_real_id))
        message = Message(MyMessage.MSG_TYPE_C2S_CLIENT_STATUS, self.client_real_id, receive_id)
        sys_name = platform.system()
        if sys_name == "Darwin":
            sys_name = "Mac"
        # Debug for simulation mobile system
        # sys_name = MyMessage.MSG_CLIENT_OS_ANDROID

        message.add_params(MyMessage.MSG_ARG_KEY_CLIENT_STATUS, status)
        message.add_params(MyMessage.MSG_ARG_KEY_CLIENT_OS, sys_name)

        if hasattr(self.args, "using_mlops") and self.args.using_mlops and \
                status == ClientMasterManager.RUN_FINISHED_STATUS_FLAG:
            mlops.log_server_payload(self.args.run_id, self.client_real_id, json.dumps(message.get_params()))
        else:
            self.send_message(message)

    def report_training_status(self, status):
        """
        Report the training status to MLOps.

        Args:
            status: Training status to be reported.

        Note:
            This method logs the training status using MLOps.

        """
        mlops.log_training_status(status)

    def sync_process_group(self, round_idx, model_params=None, client_index=None, src=0):
        """
        Synchronize the process group with information about the current training round.

        Args:
            round_idx: The current training round index.
            model_params: Model parameters (default is None).
            client_index: Client index (default is None).
            src: Source of the synchronization (default is 0).

        Note:
            This method broadcasts information about the current training round to the process group.

        """
        logging.info("sending round number to pg")
        round_number = [round_idx, model_params, client_index]
        dist.broadcast_object_list(
            round_number, src=src, group=self.trainer_dist_adapter.process_group_manager.get_process_group(),
        )
        logging.info("round number %d broadcast to process group" % round_number[0])

    def __train(self):
        """
        Perform the training for the current round.

        Note:
            This method initiates the training process and sends the updated model to the server.

        """
        logging.info("#######training########### round_id = %d" % self.round_idx)

        mlops.event("train", event_started=True, event_value=str(self.round_idx))

        weights, local_sample_num = self.trainer_dist_adapter.train(self.round_idx)

        mlops.event("train", event_started=False, event_value=str(self.round_idx))

        # the current model is still DDP-wrapped under cross-silo-hi setting
        if self.args.scenario == FEDML_CROSS_SILO_SCENARIO_HIERARCHICAL:
            weights = convert_model_params_from_ddp(weights)

        self.send_model_to_server(0, weights, local_sample_num)

    def run(self):
        super().run()
